### ISSUE TITLE
[BUG]: Parameter `model.norm.weight` failed at the gradient reduction. 

### ISSUE THREAD

**Issue by fancyerii** [hpcaitech/ColossalAI#4984]
### 🐛 Describe the bug

I am running the Colossal-LLaMA-2 example. I can run it with zero2_cpu. But when I switch to gemini or gemini_auto, it crashed with:

```
loss.backward()
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 479, in backward
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/booster/booster.py", line 159, in backward
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 479, in backward
            loss.backward()return handle_torch_function(return handle_torch_function(


  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/overrides.py", line 1534, in handle_torch_function
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/overrides.py", line 1534, in handle_torch_function
      File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 479, in backward
optimizer.backward(loss)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/zero/gemini/gemini_optimizer.py", line 265, in back
ward
    return handle_torch_function(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/overrides.py", line 1534, in handle_torch_function
    return handle_torch_function(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/overrides.py", line 1534, in handle_torch_function
    self.module.backward(loss)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/zero/gemini/gemini_ddp.py", line 305, in backward
    return handle_torch_function(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/overrides.py", line 1534, in handle_torch_function
    loss.backward()
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 479, in backward
    result = torch_func_method(public_api, types, args, kwargs)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/tensor/colo_tensor.py", line 81, in __torch_functio
n__
    result = torch_func_method(public_api, types, args, kwargs)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/tensor/colo_tensor.py", line 81, in __torch_functio
n__
    return backward_tensor.backward(**tensor_kwargs)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 488, in backward
    return backward_tensor.backward(**tensor_kwargs)
      File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 488, in backward
return handle_torch_function(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/overrides.py", line 1534, in handle_torch_function
    result = torch_func_method(public_api, types, args, kwargs)
      File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/tensor/colo_tensor.py", line 81, in __torch_function__
result = torch_func_method(public_api, types, args, kwargs)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/tensor/colo_tensor.py", line 81, in __torch_function__
    result = torch_func_method(public_api, types, args, kwargs)
          File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/tensor/colo_tensor.py", line 81, in __torch_function__
return backward_tensor.backward(**tensor_kwargs)result = torch_func_method(public_api, types, args, kwargs)

  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 488, in backward
return backward_tensor.backward(**tensor_kwargs)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/tensor/colo_tensor.py", line 81, in __torch_function__
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 488, in backward
    torch.autograd.backward(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/autograd/__init__.py", line 197, in backward
    torch.autograd.backward(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/autograd/__init__.py", line 197, in backward
    return backward_tensor.backward(**tensor_kwargs)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 488, in backward
    return backward_tensor.backward(**tensor_kwargs)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 488, in backward
    result = torch_func_method(public_api, types, args, kwargs)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/tensor/colo_tensor.py", line 81, in __torch_functio
n__
        Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward passVariable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass

  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/zero/gemini/gemini_ddp.py", line 320, in grad_handle
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/zero/gemini/gemini_ddp.py", line 320, in grad_handle
    return backward_tensor.backward(**tensor_kwargs)
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/_tensor.py", line 488, in backward
    torch.autograd.backward(
      File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/autograd/__init__.py", line 197, in backward
torch.autograd.backward(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/autograd/__init__.py", line 197, in backward
    torch.autograd.backward(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/autograd/__init__.py", line 197, in backward
    torch.autograd.backward(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/autograd/__init__.py", line 197, in backward
    raise RuntimeError(
raise RuntimeError(
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass    RuntimeError
Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward passRuntimeError:
:   File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/zero/gemini/gemini_ddp.py", line 320, in grad_handle
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/zero/gemini/gemini_ddp.py", line 320, in grad_handle
Parameter `model.norm.weight` failed at the gradient reduction. Some unsupported torch function is operated upon this parameter.Parameter `model.norm.weight` failed at the gradient reduction. Some unsupported torch function is operated upon this parameter.

    torch.autograd.backward(
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/torch/autograd/__init__.py", line 197, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/home/ubuntu/miniconda3/envs/py39_torch113/lib/python3.9/site-packages/colossalai/zero/gemini/gemini_ddp.py", line 320, in grad_handle
```

My launch script:

```
colossalai run --nproc_per_node 8   train.py \
    --pretrained /nas/lili/models_hf/13B-chat \
    --dataset "/nas/lili/colossalaitest/Colossal-LLaMA-2/spliced_tokenized_output_arrow/part-00000" \
    --plugin "gemini" \
    --save_interval 400 \
    --save_dir "13b_rv" \
    --tensorboard_dir tbdir \
    --config_file config-13b.json \
    --num_epochs 1 \
    --micro_batch_size 1 \
    --lr 1e-4 \
    --mixed_precision "bf16" \
    --grad_clip 1.0 \
    --weight_decay 0.01 \
    --warmup_steps 100 \
    --use_grad_checkpoint \
    --use_flash_attn
```

### Environment

cuda 11.7
cudnn 8.9.4.25_cuda11
python 3.9
pytorch 1.13.1+cu117
colossal 0.3.3

**Comment by JThh** [comment]
I cannot see the exact issue from your description. But the error should be raised at [line](https://github.com/hpcaitech/ColossalAI/blob/4e4a10c97d2bb485c148e727400052f9efe69864/colossalai/zero/gemini/gemini_ddp.py#L331).

**Comment by JThh** [comment]
Would you please let me know your `config` file?

**Comment by fancyerii** [comment]
@JThh I don't use any config file. I just start it with:
```
colossalai run --nproc_per_node 8   train.py \
    --pretrained /nas/lili/models_hf/13B-chat \
    --dataset "/nas/lili/colossalaitest/Colossal-LLaMA-2/spliced_tokenized_output_arrow/part-00000" \
    --plugin "gemini" \
    --save_interval 400 \
    --save_dir "13b_rv" \
    --tensorboard_dir tbdir \
    --config_file config-13b.json \
    --num_epochs 1 \
    --micro_batch_size 1 \
    --lr 1e-4 \
    --mixed_precision "bf16" \
    --grad_clip 1.0 \
    --weight_decay 0.01 \
    --warmup_steps 100 \
    --use_grad_checkpoint \
    --use_flash_attn
```

**Comment by fancyerii** [comment]
Anyone could help? I search [here](https://github.com/hpcaitech/ColossalAI/issues?q=unsupported+torch+function) and find many similar problems without answer.

I also tried [llama2 example](https://github.com/hpcaitech/ColossalAI/tree/main/examples/language/llama2). Because [this issue](https://github.com/huggingface/transformers/issues/26755), I have to use transformers 4.33.3 and I can run script/gemini_auto.sh with 70b correctly.

So the difference between examples/language/llama2 and application/Colossal-LLaMA-2 may be related to the bug.
And I also use transformers 4.33.3 to run application/Colossal-LLaMA-2. It throws the same exception.

**Comment by Fridge003** [comment]
Hi, I found that this bug was triggered by the `rms_norm` module of flash attention.
You can try commenting the last two lines of function `replace_with_flash_attention` in `colossalai_llama2/utils/flash_attention_patch.py` in the following way:
```python
def replace_with_flash_attention(model: LlamaForCausalLM) -> None:
    for name, module in model.named_modules():
        if isinstance(module, LlamaAttention):
            module.forward = MethodType(attention_forward, module)
        if isinstance(module, LlamaModel):
            module._prepare_decoder_attention_mask = MethodType(_prepare_decoder_attention_mask, module)
       #  if isinstance(module, LlamaRMSNorm):
       #    module.forward = MethodType(rms_norm_forward, module)
```
 and the bug might be fixed.

**Comment by fancyerii** [comment]
Thanks. I also find the problem and use replace_with_flash_attention. It runs correctly now. But I have to switch back to transformers==4.33.3. Hope you can fix it quickly.

**Comment by fancyerii** [comment]
Thanks. I also found the problem and tested replace_xformers. It runs correctly now. But I have to switch back to transformers==4.33.3. 
It's a workaround to comment that two lines or permanent solution?

**Comment by Fridge003** [comment]
> Thanks. I also found the problem and tested replace_xformers. It runs correctly now. But I have to switch back to transformers==4.33.3.
> It's a workaround to comment that two lines or permanent solution?

Currently it's just a workaround, we are still finding a better solution. Thank you for your issue!

**Comment by fancyerii** [comment]
By the way, what's the difference of [replace_xformers](https://github.com/hpcaitech/ColossalAI/blob/main/examples/language/llama2/attn.py) and replace_with_flash_attention?

**Comment by Fridge003** [comment]
> By the way, what's the difference of [replace_xformers](https://github.com/hpcaitech/ColossalAI/blob/main/examples/language/llama2/attn.py) and replace_with_flash_attention?

`replace_xformers` use xformers's version of flash attention, while `replace_with_flash_attention` use the original version developed by Tri Dao. Also `replace_xformers` doesn't include replacing the rms_norm, so it will not trigger this bug.

**Comment by fancyerii** [comment]
Thanks.

**Comment by fancyerii** [comment]
@Fridge003 if LlamaRMSNorm is not used, then no RMSNorm is used? Will the training be correct?

**Comment by Fridge003** [comment]
> @Fridge003 if LlamaRMSNorm is not used, then no RMSNorm is used? Will the training be correct?

LlamaRMSNorm is still used, but not replaced to the flash attention version. The training will be correct, but a little bit slower.

**Comment by fancyerii** [comment]
I got it. So when I use other plugin such as zero_cpu, I can still use the original codes. If I want to use gemini or gemini_auto, then I need comment out them.

**Comment by Fridge003** [comment]
> I got it. So when I use other plugin such as zero_cpu, I can still use the original codes. If I want to use gemini or gemini_auto, then I need comment out them.

Currently yes, we will try to fix this later.
